src/Database/Redis/Connection.hs

```diff
@@ -215,38 +229,69 @@ connectCluster bootstrapConnInfo = do
+                    connectTLSParams = connectTLSParams',
+                    connectAuth      = connectAuth',
+                    idleTime         = connectMaxIdleTime bootstrapConnInfo,
+                    maxResources     = connectMaxConnections bootstrapConnInfo,
+                    timeoutOpt       = timeoutOptUs
+                }
+                withAuth = tcpConnWithAuth connectAuth' connectTLSParams'
+                clusterConnection = Cluster.connect withAuth infos shardMapVar isConnectionReadOnly refreshShardMapWithNodeConn tcpInfo
+            -- pool <- createPool (clusterConnect isConnectionReadOnly clusterConnection) Cluster.disconnect 3 (connectMaxIdleTime bootstrapConnInfo) (connectMaxConnections bootstrapConnInfo)
+            connection <- clusterConnect isConnectionReadOnly clusterConnection
+            return $ ClusteredConnection shardMapVar connection
     where
-      withAuth host port timeout = do
-        conn <- PP.connect host port timeout
-        conn' <- case connectTLSParams bootstrapConnInfo of
-                  Nothing -> return conn
-                  Just tlsParams -> PP.enableTLS tlsParams conn
-        PP.beginReceiving conn'
-
-        runRedisInternal conn' $ do
-            -- AUTH
-            case connectAuth bootstrapConnInfo of
-                Nothing   -> return ()
-                Just pass -> do
-                  resp <- auth pass
-                  case resp of
-                    Left r -> liftIO $ throwIO $ ConnectAuthError r
-                    _      -> return ()
-        return $ PP.toCtx conn'
 
       clusterConnect :: Bool -> IO Cluster.Connection -> IO Cluster.Connection
       clusterConnect readOnlyConnection connection = do
-          clusterConn@(Cluster.Connection nodeMap _ _ _ _) <- connection
-          nodesConns <-  sequence $ ( PP.fromCtx . (\(Cluster.NodeConnection ctx _ _) -> ctx ) . snd) <$> (HM.toList nodeMap)
-          when readOnlyConnection $
-                  mapM_ (\conn -> do
-                          PP.beginReceiving conn
-                          runRedisInternal conn readOnly
-                      ) nodesConns
+          clusterConn@(Cluster.Connection nodeMapVar _ _ _ _) <- connection
+          nodeMap <- readMVar nodeMapVar
+          let nodeList = HM.toList nodeMap
+          maybeConns <- sequence $ (ctxToConn . snd) <$> nodeList
+          let nodeConnsPair = zip maybeConns nodeList
+          workingNodes <- mapM (\(maybeConn, nodeConn) -> case maybeConn of
+                    Just conn -> do
+                        when readOnlyConnection $ do
+                            PP.beginReceiving conn
+                            runRedisInternal conn readOnly >> return ()
+                        return $ Just nodeConn
+                    Nothing -> return Nothing
+                ) nodeConnsPair
+          let newMap = HM.fromList $ catMaybes workingNodes
+          _ <- forkIO $ modifyMVar_ nodeMapVar (\_ -> return newMap)
```

Comment:
Why forkIO?

Reply:
@neeraj97  I don't remember exactly but I think this could be amongst the changes that I was trying for fixing that deadlock in refreshShardMap. I removed this now